File: 124.二叉树中的最大路径和.cc
#include <algorithm>

using std::max;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int final_ans = INT_MIN;
        maxPath(root, final_ans);
        return final_ans;
    }

    int maxPath(TreeNode* root, int& final_ans) {
        if (root == nullptr)
            return 0;
        int left = max(maxPath(root->left, final_ans), 0);
        int right = max(maxPath(root->right, final_ans), 0);
        int newPath = root->val + left + right;
        final_ans = max(final_ans, newPath);
        return root->val + max(left, right);
    }
};

File: Python/124.二叉树中的最大路径和.py
from typing import Optional
import sys

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def __init__(self):
        self.ans: int = -sys.maxsize - 1

    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.maxPath(root)
        return self.ans

    def maxPath(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        left: int = max(0, self.maxPath(root.left))
        right: int = max(0, self.maxPath(root.right))
        newPath: int = root.val + left + right
        self.ans = max(self.ans, newPath)
        return root.val + max(left, right)
